Add optional label customization to TimePicker

TodorGinchev opened the pull request:
Add optional custom localization for TimePicker, the same way DatePicker has it
This addresses issue 84196

Similarly to DatePicker component, this commit allows TimePicker to have custom localization for hourLabel, minuteLabel and invalidTimeLabel.

This fixed the following issue: https://github.com/flutter/flutter/issues/84196

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

TodorGinchev committed:
Add optional custom localization for TimePicker, the same way DatePicker has it
This addresses issue 84196

TodorGinchev commented:
@googlebot I fixed it.

TodorGinchev commented:
@googlebot I fixed it.

mateusfccp commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1304,6 +1307,21 @@ class _TimePickerInput extends StatefulWidget {
 
+  /// Optionally provide your own validation error text.
+  final String? errorInvalidText;
+  ///
+  /// If null, the button uses [MaterialLocalizations.invalidTimeLabel].
```
```suggestion
```

mateusfccp commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1304,6 +1307,21 @@ class _TimePickerInput extends StatefulWidget {
+
+  /// Optionally provide your own hour label text.
+  final String? hourLabelText;
+  ///
+  /// If null, the button uses [MaterialLocalizations.timePickerHourLabel].
```
```suggestion
```

mateusfccp commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1304,6 +1307,21 @@ class _TimePickerInput extends StatefulWidget {
+
+  /// Optionally provide your own minute label text.
+  final String? minuteLabelText;
+  ///
+  /// If null, the button uses [MaterialLocalizations.timePickerMinuteLabel].
```
```suggestion
```

mateusfccp commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1808,6 +1835,21 @@ class TimePickerDialog extends StatefulWidget {
 
+  /// Optionally provide your own validation error text.
+  final String? errorInvalidText;
+  ///
+  /// If null, the button uses [MaterialLocalizations.invalidTimeLabel].
```
```suggestion
```

mateusfccp commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1808,6 +1835,21 @@ class TimePickerDialog extends StatefulWidget {
+
+  /// Optionally provide your own hour label text.
+  final String? hourLabelText;
+  ///
+  /// If null, the button uses [MaterialLocalizations.timePickerHourLabel].
```
```suggestion
```

mateusfccp commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -1808,6 +1835,21 @@ class TimePickerDialog extends StatefulWidget {
+
+  /// Optionally provide your own minute label text.
+  final String? minuteLabelText;
+  ///
+  /// If null, the button uses [MaterialLocalizations.timePickerMinuteLabel].
```
```suggestion
```

mateusfccp commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -861,6 +861,14 @@ void _testsInput() {
     expect(find.text(helpText), findsOneWidget);
   });
 
+  testWidgets('Hour and minute label text is used - Input', (WidgetTester tester) async {
+    const String hourLabelText = 'Custom hour label';
+    const String minuteLabelText = 'Custom minute label';
+    await mediaQueryBoilerplate(tester, true, entryMode: TimePickerEntryMode.input, hourLabelText: hourLabelText, minuteLabelText: minuteLabelText );
+    expect(find.text(hourLabelText), findsOneWidget);
+    expect(find.text(minuteLabelText), findsOneWidget);
+  });
```
I think we could make this two tests instead of one, so we may detect individual regressions.

TodorGinchev replied:
Done, thank you for the feedback

mateusfccp reviewed (commented):
Some initial opinion on this.

About the docstrings, I removed some of them because they both wouldn't work (as it is below the item it intends to document) and is different from the already existent pattern (see, for instance, `helpText`,  which doesn't include null case explanation).

shihaohong commented on packages/flutter/lib/src/material/time_picker.dart:
```diff
@@ -2342,6 +2375,9 @@ Future<TimeOfDay?> showTimePicker({
   String? cancelText,
   String? confirmText,
   String? helpText,
+  String? errorInvalidText,
```
Can you add some documentation for these parameters above? I'm thinking line 2322 where this currently exists: 
```dart
/// Optional strings for the [helpText], [cancelText], and [confirmText] can be
/// provided to override the default values.
```

TodorGinchev replied:
Done

shihaohong commented on packages/flutter/test/material/time_picker_test.dart:
```diff
@@ -861,6 +861,19 @@ void _testsInput() {
     expect(find.text(helpText), findsOneWidget);
   });
 
+  testWidgets('Hour label text is used - Input', (WidgetTester tester) async {
```
Please add a test for the invalid error text as well

TodorGinchev replied:
Done

shihaohong reviewed (approved):
LGTM! Although it looks like some of the presubmit checks are failing but don't seem related to your changes. Might require a rebase to upstream, but it could also be an issue with Cirrus. I'll investigate it for a bit.

TodorGinchev commented:
> LGTM! Although it looks like some of the presubmit checks are failing but don't seem related to your changes. Might require a rebase to upstream, but it could also be an issue with Cirrus. I'll investigate it for a bit.

Thanks! I tried to re-trigger several times during the day but no luck.

shihaohong commented:
Could you try rebasing your branch with master? That might fix the problem. Some of the newer PRs seem to not have the Cirrus failures anymore

TodorGinchev commented:
> Could you try rebasing your branch with master? That might fix the problem. Some of the newer PRs seem to not have the Cirrus failures anymore

I rebased, issue still persists, I do not really know how to fix this apart from re-triggering the CI, which fails immediately with 400 "Invalid JWT Signature."

shihaohong commented:
Are you sure you rebased? I'm not seeing the commit in the PR history. Could you check and see if you pushed your rebased commits onto GitHub

TodorGinchev commented:
> > Could you try rebasing your branch with master? That might fix the problem. Some of the newer PRs seem to not have the Cirrus failures anymore
> 
> I rebased, issue still persists, I do not really know how to fix this apart from re-triggering the CI, which fails immediately with 400 "Invalid JWT Signature."

@shi

> Are you sure you rebased? I'm not seeing the commit in the PR history. Could you check and see if you pushed your rebased commits onto GitHub

You are completely right, I rebased my forked master, not this branch :D

mateusfccp commented:
@TodorGinchev It seems you are not very acquainted with git. Rebasing is different from merging.

https://www.atlassian.com/git/tutorials/merging-vs-rebasing

I suggest you to study a little about this, as we usually rebase before merging into master.

shihaohong commented:
@mateusfccp For Flutter, we squash and merge PRs so the history cleanliness doesn't quite matter, unless you're doing it for your own personal sake :P

mateusfccp commented:
I see @shihaohong , this make sense
